onboard/catkin_ws/src/controls/src/controls.cpp

```diff
@@ -0,0 +1,522 @@
+#include "controls.h"
+
+#include <custom_msgs/ControlTypes.h>
+#include <custom_msgs/PIDAxesInfo.h>
+#include <custom_msgs/PIDGain.h>
+#include <custom_msgs/PIDGains.h>
+#include <custom_msgs/SetControlTypes.h>
+#include <custom_msgs/SetPIDGains.h>
+#include <custom_msgs/SetPowerScaleFactor.h>
+#include <custom_msgs/SetStaticPower.h>
+#include <custom_msgs/ThrusterAllocs.h>
+#include <geometry_msgs/Pose.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/TransformStamped.h>
+#include <geometry_msgs/Twist.h>
+#include <geometry_msgs/Vector3.h>
+#include <geometry_msgs/Vector3Stamped.h>
+#include <nav_msgs/Odometry.h>
+#include <ros/ros.h>
+#include <std_msgs/Bool.h>
+#include <std_msgs/Float64.h>
+#include <std_srvs/SetBool.h>
+#include <std_srvs/Trigger.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2/LinearMath/Vector3.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_ros/transform_listener.h>
+
+#include <Eigen/Dense>
+#include <array>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include "controls_types.h"
+#include "controls_utils.h"
+#include "pid_manager.h"
+#include "thruster_allocator.h"
+
+Controls::Controls(int argc, char **argv, ros::NodeHandle &nh, std::unique_ptr<tf2_ros::Buffer> tf_buffer) {
+    // Get parameters from launch file
+    ros::param::get("~sim", sim);
+    ros::param::get("~enable_position_pid", enable_position_pid);
+    ros::param::get("~enable_velocity_pid", enable_velocity_pid);
+    ros::param::get("~cascaded_pid", cascaded_pid);
+
+    // Store transform buffer
+    this->tf_buffer = std::move(tf_buffer);
+
+    // Initialize desired position to have valid orientation
+    desired_position.orientation.w = 1.0;
+
+    // Use desired position as the default control type for all axes
+    ControlsUtils::populate_axes_map<ControlTypesEnum>(ControlTypesEnum::DESIRED_POSITION, control_types);
+
+    // Get paths to wrench matrix and its pseudoinverse from robot config file
```

Comment:
For the config variables, why are some initialized here while others are initialized in 'controls.h`?

Reply:
Good question. It is a good practice to initialize variables in the cpp file, not in the header file. I have made this change.